File: backend/common-service/src/main/java/com/familring/common_service/dto/BaseResponse.java
package com.familring.common_service.dto;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Getter
@AllArgsConstructor
@NoArgsConstructor
public class BaseResponse<T> {
    private int statusCode;
    private String message;
    private T data;

    public static <T> BaseResponse<T> create(int statusCode, String message) {
        return new BaseResponse<>(statusCode, message, null);
    }

    public static <T> BaseResponse<T> create(int statusCode, String message, T dto) {
        return new BaseResponse<>(statusCode, message, dto);
    }
}

File: android/data/src/main/java/com/familring/data/network/response/BaseResponse.kt
package com.familring.data.network.response

import com.familring.data.exception.ApiException
import com.familring.data.exception.RefreshTokenExpiredException
import com.familring.domain.model.ApiResponse
import java.io.IOException

data class BaseResponse<T>(
    val statusCode: Int,
    val data: T?,
    val message: String,
)

suspend fun <T> emitApiResponse(
    apiResponse: suspend () -> BaseResponse<T>,
    default: T,
): ApiResponse<T> =
    runCatching {
        apiResponse()
    }.fold(
        onSuccess = { result ->
            ApiResponse.Success(data = result.data ?: default)
        },
        onFailure = { e ->
            when (e) {
                is ApiException ->
                    ApiResponse.Error.ServerError(
                        code = e.error.errorCode ?: "500",
                        message = e.error.errorMessage ?: "서버 내부 오류가 발생했습니다.",
                    )

                is RefreshTokenExpiredException ->
                    ApiResponse.Error.TokenError(
                        code = e.error.errorCode ?: "500",
                        message = e.error.errorMessage ?: "서버 내부 오류가 발생했습니다.",
                    )

                is IOException ->
                    ApiResponse.Error.NetworkError(
                        message = e.message ?: "Network Error",
                    )

                else ->
                    ApiResponse.Error.UnknownError(
                        message = e.message ?: "Unknown Error",
                    )
            }
        },
    )
